### ISSUE TITLE
[Question] Custom Database Connections for Tenants

### ISSUE THREAD

**Issue by johannesschobel** [hyn/multi-tenant#557]
Dear @luceos and community,

i recently toyed around with hyn, and i was able to get it to work - even though i have a quite complex application architecture (e.g., see another issue posted). That was a really (!) awesome experience 👍 

However, i have an additional question - i was not able to find respective information on the docs and i am not sure if this is even possible:

Imagine the following scenario:
I want to use the "each customer has its own database" approach, because i really like to separate the data on "database" layer. i am fully aware that this, however, may cause issues related to hosting such applications as most "common hosters" do not offer such features as "nearly unlimited databases" or "automatically creating new databases" on the fly.

And this is, where the shit hits the fan 😆 

Basically, what i want to do is:
1) Create the database via a web-form provided by my web-hoster and set everything up. Note that my hoster automatically generates a random database name **i am not able to change**! Further, this database gets its own user assigned (again, i am not able to change the username!)
2) Create a new `Tenant` with hyn and append the newly created database connection parameters to this model
3) Migrate the database to this connection

Is this currently possible with hyn? Will this be possible with tenancy?
If so, how can I achieve such a setup? If not, what would be the best approach to support such a feature?

Thank you very much for your awesome work and your time!
All the best,

**Comment by luceos** [comment]
If you plan to automate interaction with the website of your hosting provider I can only recommend against it. What if the site is unreachable or they block your requests due to thinking they are spam. If they have an API that is much better though.

The easiest, and very cheap, way of running tenancy is using your own VM. If you don't want to have the burden of server set up; I can recommend https://forge.laravel.com. Digital Ocean has very cheap vm's to start with.

**Comment by johannesschobel** [comment]
wow, that was a fast resposne 😆 

sorry, i may have not provided enough information. I don't want to "automatically" create a new database on my hoster - i will do this manually. So there should / will not be a "create account and immediately start working" setup (at least not for now, haha)..

Why shouldn't i use such an approach? Can you provide some details in order to reproduce your though-process on this?

Thank you very much!

**Comment by luceos** [comment]
Aha. The process is possible in both packages. Let's stick to the stable one though.

You can configure your own database configuration whenever tenancy needs it. For that to work you can listen to a specific event ; hyn tenancy events database configuration loading, sorry on mobile :v:

As to my thought process. This package was developed to give you as much automation as possible, hence the vhost configuration files for apache and nginx, but also the database creation and deletion.. Not using those features is like buying a tesla and retrofitting it with a diesel engine. :relaxed:

**Comment by johannesschobel** [comment]
haha, that is  a good example..
however, i would like to have some kind of "the admin can move the tenant ot another server" functionality.. 

For example, `TenantA` may be a "premium" user that should get a more "powerful" (i.e., less users per server) machine.. So i would like to move his database from `ip-address-of-slow-server` to `ip-address-of-powerful-server`. This way, however, the admin must "re-configure" the database-connection properties (e.g., address of the db, port, username, password, ...). Maybe one could also use some kind of "you can have a postgreSQL database instead of mySQL" approach.. 

Further, i did not find any "database fields" where such connection parameters of a tenant can be stored. From what i read, these database-connection properties are "calculated" from the tenant object (i.e., it generates a random "database name", creates a username with the same name, creates a hash for the pw from various attributes, ...)

**Comment by luceos** [comment]
- Migrating to another server; as the code base doesn't change (and you can rely on block storage for most other files) we only have to offer a way to reconfigure another connection that is being seen as the connection on which to create users and databases. This feature exists under the name ["managing connection"](https://laravel-tenancy.com/docs/hyn/5.2/repositories#managing-connection). The default system connection still holds information on hostnames and websites, but the managing connection is responsible for maintaining the database and database user.
- There are no database fields. The database connection configuration is generated dynamically, however you can still specify your own configuration before hand, as pointed out yesterday. To do so:

Listen to the even `Hyn\Tenancy\Events\Database\ConfigurationLoading` you can see how it is used here:

https://github.com/hyn/multi-tenant/blob/73d7d6d9ff746603987687f3e8f31f0df7e04519/src/Database/Connection.php#L296-L327

You'll achieve most flexibility by modifying the `$mode` to get the value `Hyn\Tenancy\Database\Connection::DIVISION_MODE_BYPASS` and setting the `$configuration` value to whatever array you need.

**Comment by johannesschobel** [comment]
Hey @luceos ,
sorry for the delay, but at the moment I am quite busy with finishing my PhD thesis :joy: 

If I would use the `managed by database connection` feature, however, I would need to specify a new `connection` in the `config/database.php` configuration file, right? So, basically, for each tenant, i would need to add such a block?!

Further, if i want to store the db-connection information (e.g., the host to connect to, db-name, db-user, ...) in the `system database` for the `Website`, i would need to extend the current models (and db-schema), right?

Thank you so much for your awesome support!

**Comment by luceos** [comment]
> however, I would need to specify a new connection in the config/database.php configuration file, right? So, basically, for each tenant, i would need to add such a block?!

If you need a separate database __server__ per tenant then you need a separate database connection set up per tenant. You can however do that during runtime as well! A possible location would be the `ConfigurationLoading` event, so that the system connection is set up before the Tenant connection..

> Further, if i want to store the db-connection information (e.g., the host to connect to, db-name, db-user, ...) in the system database for the Website, i would need to extend the current models (and db-schema), right?

That would be possible. You can easily extend the implementation and configure it in the tenancy.php.

Good luck with your thesis!

**Comment by SebastianSchoeps** [comment]
For any one who stumbles over this: if you need to access the website object in the listener to `ConfigurationLoading` you could cache the website object in a listener to `Events\Websites\Identified`. Then you can access it in your listener.

@luceos Would you accept an PR which hands the website object to the `ConfigurationLoading`event?
```$this->emitEvent(new Events\Database\ConfigurationLoading($mode, $clone, $this));```
would become
```$this->emitEvent(new Events\Database\ConfigurationLoading($mode, $clone, $this, $website));```

**Comment by luceos** [comment]
Yes, seems like the original question was answered. Let's not trail into having a discussion on different topics inside one issue.
